### PR TITLE
p2p class sync tests

### PR DESCRIPTION
**Pull request by CHr15F0x**
Part of #1944 .

### CODE REVIEW

File: crates/pathfinder/src/sync/checkpoint.rs [hunk 1]
```diff
@@ -448,13 +438,35 @@ impl Sync {
...
+    class_stream
+        .map_err(Into::into)
+        .map_ok(class_definitions::verify_layout)
+        .and_then(std::future::ready)
+        // Block numbers in the stream are guaranteed to be correct but we need to check that
+        // the streamed classes belong to the right block.
+        .and_then(|x| class_definitions::verify_declared_at(storage.clone(), x))
+        .and_then(class_definitions::verify_hash)
+        .try_chunks(1000)
+        .map_err(|e| e.1)
```
[the remark this answers is not in the document]
Comment (CHr15F0x) [reply crates/pathfinder/src/sync/checkpoint.rs:460]: Yes, this is definitely too many.

File: crates/pathfinder/src/sync/checkpoint.rs [hunk 2]
```diff
@@ -448,13 +438,35 @@ impl Sync {
...
 
-async fn handle_events_stream(
+async fn handle_class_stream(
+    class_stream: impl futures::Stream<Item = Result<PeerData<Class>, anyhow::Error>>,
+    storage: Storage,
+) -> Result<(), SyncError> {
+    class_stream
+        .map_err(Into::into)
+        .map_ok(class_definitions::verify_layout)
+        .and_then(std::future::ready)
```
[the remark this answers is not in the document]
Comment (CHr15F0x) [reply crates/pathfinder/src/sync/checkpoint.rs:454]: This is an adapter for the previous stage returning `Result<>`. I'll make `verify_layout` async instead.

File: crates/pathfinder/src/sync/checkpoint.rs [hunk 3]
```diff
@@ -448,12 +454,20 @@ async fn handle_class_stream(
     class_stream: impl futures::Stream<Item = Result<PeerData<Class>, anyhow::Error>>,
     storage: Storage,
 ) -> Result<(), SyncError> {
+    let mut expected_classes_for_block = Arc::new(Mutex::new(HashSet::new()));
```
Comment (CHr15F0x) [inline crates/pathfinder/src/sync/checkpoint.rs:457]: This is due to `scan()` limitations.

File: crates/pathfinder/src/sync/checkpoint.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (Mirko-von-Leipzig) [reply crates/pathfinder/src/sync/checkpoint.rs]: Kinda cool that you can just do that though. But yeah maybe just async it instead.

File: crates/pathfinder/src/sync/class_definitions.rs [hunk 4]
```diff
@@ -139,6 +151,47 @@ pub(super) fn verify_layout(
...
 }
 
+/// Checks if the streamed class is expected to be declared at the given block.
+/// If yes, it removes its hash from the "expected" set and returns the class, otherwise it returns an error.
+/// When the set of expected classes for the block is exhausted it fetches the set for the next block.
+///
+/// This function relies on the guarantee that the block numbers in the stream are correct.
+pub(super) async fn verify_declared_at(
+    storage: Storage,
+    expected_classes_for_block: Arc<Mutex<HashSet<ClassHash>>>,
```
Comment (Mirko-von-Leipzig) [inline crates/pathfinder/src/sync/class_definitions.rs:161]: An alternative to consider: maybe use a `impl Stream<HashSet<ClassHash>>`? Then we spawn a task doing the storage thingy in a loop? Simplifies this code here by "moving" the db stuff outside which lets this function logic focus on only checking.

File: crates/pathfinder/src/sync/class_definitions.rs [hunk 5]
```diff
@@ -151,44 +153,72 @@ pub(super) async fn verify_layout(
...
+            yield declared_at_this_block;
+
+            start += 1;
+        }
     }
+}
+
+/// This function relies on the guarantee that the block numbers in the stream are correct.
+pub(super) fn verify_declared_at(
+    storage: Storage,
```
Comment (Mirko-von-Leipzig) [inline crates/pathfinder/src/sync/class_definitions.rs:191]:
```suggestion
```

**Review by Mirko-von-Leipzig — APPROVED** [review]

**Pull request merged**